golem-component-service-base/src/repo/component.rs

```diff
@@ -0,0 +1,467 @@
+            .bind(namespace)
+            .bind(component_id)
+            .fetch_all(self.db_pool.deref())
+            .await
+            .map_err(|e| e.into())
+    }
+
+    async fn get_all(&self, namespace: &str) -> Result<Vec<ComponentRecord>, RepoError> {
+        sqlx::query_as::<_, ComponentRecord>("SELECT namespace, component_id, version, name, size, user_component, protected_component, protector_version, metadata FROM components WHERE namespace = $1")
+            .bind(namespace)
+            .fetch_all(self.db_pool.deref())
+            .await
+            .map_err(|e| e.into())
+    }
+
+    async fn get_latest_version(
+        &self,
+        namespace: &str,
+        component_id: &Uuid,
+    ) -> Result<Option<ComponentRecord>, RepoError> {
+        sqlx::query_as::<_, ComponentRecord>(
+            "SELECT namespace, component_id, version, name, size, user_component, protected_component, protector_version, metadata FROM components WHERE namespace = $1 AND component_id = $2 ORDER BY version DESC LIMIT 1",
+        ).bind(namespace)
+            .bind(component_id)
+            .fetch_optional(self.db_pool.deref())
+            .await
+            .map_err(|e| e.into())
+    }
+
+    async fn get_by_version(
+        &self,
+        namespace: &str,
+        component_id: &Uuid,
+        version: u64,
+    ) -> Result<Option<ComponentRecord>, RepoError> {
+        sqlx::query_as::<_, ComponentRecord>(
+            "SELECT namespace, component_id, version, name, size, user_component, protected_component, protector_version, metadata FROM components WHERE namespace = $1 AND component_id = $2 AND version = $3",
+        )
+            .bind(namespace)
+            .bind(component_id)
+            .bind(version as i64)
+            .fetch_optional(self.db_pool.deref())
+            .await
+            .map_err(|e| e.into())
+    }
+
+    async fn get_by_name(
+        &self,
+        namespace: &str,
+        name: &str,
+    ) -> Result<Vec<ComponentRecord>, RepoError> {
+        sqlx::query_as::<_, ComponentRecord>(
+            "SELECT namespace, component_id, version, name, size, user_component, protected_component, protector_version, metadata FROM components WHERE namespace = $1 AND name = $2",
+        )
+            .bind(namespace)
+            .bind(name)
+            .fetch_all(self.db_pool.deref())
+            .await
+            .map_err(|e| e.into())
+    }
+
+    async fn get_ids_by_name(&self, namespace: &str, name: &str) -> Result<Vec<Uuid>, RepoError> {
+        let result = sqlx::query(
+            "SELECT distinct component_id as component_id FROM components WHERE namespace = $1 AND name = $2"
+        )
+            .bind(namespace)
+            .bind(name)
+            .fetch_all(self.db_pool.deref())
+            .await?;
+
+        Ok(result.into_iter().map(|x| x.get("component_id")).collect())
+    }
+
+    async fn get_namespaces(&self, component_id: &Uuid) -> Result<Vec<(String, i64)>, RepoError> {
```

Comment:
What is the purpose of namespace in the result? According to the `upsert` implementation multiple namespaces is an invalid state.

Reply:
changed to  return just optional value